### ISSUE TITLE
Try to get rid of thread static structures in Parser

### ISSUE THREAD

**Issue by k-rus** [Starcounter/Starcounter#1214]
Bison-based parser uses thread static variable to keep error data. Bison already provides ability to have custom elements in the structure passed during parsing as argument. The guess is that the structure can be utilized in better way to get rid of thread static variables.

**Comment by k-rus** [comment]
@alexanderstc I guess this issue is valid and will be important when we move to DNX and 3.x, but not now. I remember we discuss similar for meta-layer. WDYT?

**Comment by alexanderstc** [comment]
I need to re-examine this.



On Wed, Jan 13, 2016 at 5:30 AM -0800, "Ruslan Fomkin" <notifications@github.com<mailto:notifications@github.com>> wrote:


@alexanderstc<https://github.com/alexanderstc> I guess this issue is valid and will be important when we move to DNX and 3.x, but not now. I remember we discuss similar for meta-layer. WDYT?

-

**Comment by k-rus** [comment]
Currently, it is backlog issue, but it seems to me we will need to fix it eventually.

**Comment by alexanderstc** [comment]
Discussed with @k-rus, probably only error reporting from the SQL parser has to be fixed (changed to use the same metalayer error buffers, probably).
Though this must be done before we call the SQL parser compatible with any thread-agnostic scheduler.

**Comment by alexanderstc** [comment]
@k-rus
> I remember we discuss similar for meta-layer.

Yes, the result was "it's OK to use thread-local vars to keep data while serving a user call, but not between user calls".
Hence we should fix SQL parser error reporting because it requires the user to call another function to retrieve an error which happened in a previous function.

**Comment by alexanderstc** [comment]
Going to add an SCSQL_ERROR * parameter to QP API functions to return the error that way instead of thread-local variable.

**Comment by alexanderstc** [comment]
Removed scsql_get_error(), passing QP error via api functions parameter by a copy.
In l0 & l1 pnext.
Internally in level0 the scerror is still a thread-local variable, but that doesn't matter now.
It can be changed to a per-context as soon as QP approach to contexts is decided.
